libs/core/algorithms/tests/unit/algorithms/util/test_is_trivially_relocatable.cpp

```diff
@@ -0,0 +1,55 @@
+//  Copyright (c) 2023 Isidoros Tsaousis-Seiras
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/algorithms/traits/is_trivially_relocatable.hpp>
+
+#include <cassert>
+
+// Trivially Copyable types are trivially relocatable
+struct TriviallyCopyable
+{
+    TriviallyCopyable(TriviallyCopyable const&) = default;
+};
+static_assert(hpx::is_trivially_relocatable_v<TriviallyCopyable>,
+    "Trivially Copyable type should be Trivially Relocatable");
+
+// Non trivially copyable types are not trivially relocatable
+// Unless they are explicitly declared as such
+struct NotTriviallyCopyable_1
+{
+    NotTriviallyCopyable_1(NotTriviallyCopyable_1 const&){};
+};
+static_assert(!hpx::is_trivially_relocatable_v<NotTriviallyCopyable_1>,
+    "Not Trivially Copyable and Not declared Trivially Relocatable type should "
+    "not be Trivially Relocatable");
+
+struct NotTriviallyCopyable_2
+{
+    NotTriviallyCopyable_2(NotTriviallyCopyable_2 const&){};
+};
+HPX_DECLARE_TRIVIALLY_RELOCATABLE(NotTriviallyCopyable_2)
+
+static_assert(hpx::is_trivially_relocatable_v<NotTriviallyCopyable_2>,
+    "Not Trivially Copyable but declared Trivially Relocatable type should "
+    "be Trivially Relocatable");
```

Comment:
Here I recommend
```
struct ExplicitlyTriviallyRelocatable1 {
    ExplicitlyTriviallyRelocatable1();
    ExplicitlyTriviallyRelocatable1(const ExplicitlyTriviallyRelocatable1&);
    ExplicitlyTriviallyRelocatable1& operator=(const ExplicitlyTriviallyRelocatable1&);
    ~ExplicitlyTriviallyRelocatable1();
};
HPX_DECLARE_TRIVIALLY_RELOCATABLE(ExplicitlyTriviallyRelocatable1);

struct ExplicitlyTriviallyRelocatable2 {
    ExplicitlyTriviallyRelocatable2();
    ExplicitlyTriviallyRelocatable2(ExplicitlyTriviallyRelocatable2&&);
    ExplicitlyTriviallyRelocatable2& operator=(ExplicitlyTriviallyRelocatable2&&);
    ~ExplicitlyTriviallyRelocatable2();
};
HPX_DECLARE_TRIVIALLY_RELOCATABLE(ExplicitlyTriviallyRelocatable2);

struct DerivedFromExplicitlyTriviallyRelocatable : ExplicitlyTriviallyRelocatable1 {
    ~DerivedFromExplicitlyTriviallyRelocatable();
};

static_assert(hpx::is_trivially_relocatable_v<ExplicitlyTriviallyRelocatable1>);
static_assert(hpx::is_trivially_relocatable_v<ExplicitlyTriviallyRelocatable2>);
static_assert(!hpx::is_trivially_relocatable_v<DerivedFromExplicitlyTriviallyRelocatable>);
```

Reply:
From my understanding, the primary difference between 1 and 2 is that 1 is only **copy** assignable/constructible while 2 is only **move** assignable/constructible, with non-trivial functions for everything. Does each structure represent a specific edge case scenario that the tests need to cover? I'm asking this to better understand the principles guiding the selection process for future test cases.

Reply:
1 is both copyable and movable; it's a classic C++98 "Rule of Three" type, which you'll probably see a lot of in real code. The fact that when you move it you get a call to the copy constructor (not a separate move constructor) doesn't really matter to me. The point is just that it's not Rule-of-Zero and it isn't move-only.
2 is move-only, exactly isomorphic to `unique_ptr`.

The point of `DerivedFromExplicitlyTriviallyRelocatable` is that back when I was first thinking about "The Best Type Traits C++ Doesn't Have," I considered the idea of opting into things like trivial relocatability and trivial comparability by using a member typedef:
```
struct S {
    using is_trivially_relocatable = void;
};
```
(The STL uses this technique for [`is_transparent`](https://stackoverflow.com/questions/20317413/what-are-transparent-comparators).) The downside to this is that then anything _derived_ from `S` _inherits_ that member typedef — i.e., anything derived from a trivially relocatable type ends up being marked as trivially relocatable — which obviously isn't going to work.

Now, neither P1144 nor HPX are actually doing anything at all with member typedefs, so we don't expect anything at all to go wrong with `DerivedFromExplicitlyTriviallyRelocatable`. But I figured it couldn't hurt to add the test case for it anyway.

Reply:
I see, both the tests and the rationale behind not using typedefs make sense, thank you :)